### PR TITLE
I/7559: Update ImageResizeUI plugin

### LINKED ISSUE #7559: Icon for ImageResizeUI plugin
We just introduced the new plugin [`ImageResizeUi`](https://github.com/ckeditor/ckeditor5/pull/7482) which allows the users to manually change the size of an image based on the given configuration.

The plugin may be displayed as a dropdown or a set of standalone buttons.

First attempt for the icon looked like:

dropdown

and 

buttons

The concept of the icon above looks OK with a dropdown but we're not sure whether we want to display the icon for every standalone button...

**The question is:**
- what icon should fit best?
- if we want to have an icon, how should we present it with standalone buttons?

**cc @oleq**

### PR DESCRIPTION
**Pull request by panr**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/contributing/git-commit-message-convention.html))

Other (image): Improved `ImageResizeUI` plugin UI. Added the dedicated icons for each option (`small`, `medium`, `large` and `original`) and the dropdown (dropdown icon).  Closes #7559.

Other (image): Improved screen reader experience for `ImageResizeUI` standalone buttons in the `image.toolbar`.

Other (core): Added icons that represent the size of the element (`object-size-*`).

### CODE REVIEW

File: packages/ckeditor5-image/src/imageresize.js [hunk 1]
```diff
@@ -60,14 +60,15 @@ export default class ImageResize extends Plugin {
 /**
  * The resize options.
  *
- * Each option should have its `name`, which is a component name definition that will be
- * used in the {@link module:image/imageresize/imageresizeui~ImageResizeUI} plugin.
- * Other properties like `label` and `value` define the following:
- * a text label for the option button and the value that will be applied to the image's width.
+ * Each option should have its own `name`, which is a component name definition that will be
```
Comment (oleq) [inline packages/ckeditor5-image/src/imageresize.js:63]: I'd rephrase it

---

```
Each option should have 2 properties:

* name: The name of the UI component registered in the global {@link module:core/editor/editorui~EditorUI#componentFactory component factory} of the editor, 
representing the button a user can click to change the size of an image,
* value: An actual image width applied when a user clicks the mentioned button 
({@link module:image/imageresize/imageresizecommand~ImageResizeCommand} gets executed). 
The value property is combined with the {@link module:image/image~ImageConfig#resizeUnit `config.image.resizeUnit`} (`%` by default), 
e.g.: `value: '50'` and `resizeUnit: '%'` give `50%`.
```

File: packages/ckeditor5-image/src/imageresize.js [hunk 2]
```diff
@@ -60,14 +60,15 @@ export default class ImageResize extends Plugin {
...
+ * used in the {@link module:image/imageresize/imageresizeui~ImageResizeUI} plugin and `value` which will
+ * be applied to the image's width during the {@link module:image/imageresize/imageresizecommand~ImageResizeCommand} execution.
  *
  * The value property is combined with the `resizeUnit` (`%` by default), eg: `value: '50'` and `resizeUnit: '%'` is `50%`.
  *
- * **NOTE:** If you want to set an option that will reset image to its original size, you need to pass a `null` value
+ * **NOTE:** There is additional mandatory property — `icon`, but it's necessary only for setting the standalone buttons configuration.
```
Comment (oleq) [inline packages/ckeditor5-image/src/imageresize.js:69]: Let's move this paragraph next to

> If you want to have separate buttons for each option \[...\]

because it's the only case when it matters. Or we may not even need this paragraph because it's already explained

> Please keep in mind that this time \*\*you should define additional \`icon\` property\*\*:

File: packages/ckeditor5-image/src/imageresize.js [hunk 3]
```diff
@@ -77,18 +78,18 @@ export default class ImageResize extends Plugin {
  *					resizeUnit: "%",
  *					resizeOptions: [ {
  *						name: 'imageResize:original',
- *						label: 'Original size',
  *						value: null
+ *						// you should add `icon` property if you're configuring the standalone buttons.
```
Comment (oleq) [inline packages/ckeditor5-image/src/imageresize.js:82]: Let's move these comments to the second part of the doc. They don't make sense if they configure a drop-down and they bring only noise.

File: packages/ckeditor5-image/src/imageresize.js [hunk 4]
```diff
@@ -124,28 +122,62 @@ export default class ImageResize extends Plugin {
  *			.then( ... )
  *			.catch( ... );
  *
- * If you want to have separate buttons for each option, pass their names instead:
+ * If you want to have separate buttons for each option, pass their names to the `image.toolbar` instead. Please keep in
```
Comment (oleq) [inline packages/ckeditor5-image/src/imageresize.js:125]: Which icons are available? How do I find them?

We should at least link to `module:image/imageresize/imageresizeui~ImageResizeOption`.

File: packages/ckeditor5-image/src/imageresize/imageresizeui.js [hunk 5]
```diff
@@ -146,33 +204,69 @@ export default class ImageResizeUI extends Plugin {
...
+	/**
+	 * A helper function for parsing resize options definitions.
+	 *
+	 * @private
+	 *
+	 * @param {Array.<module:image/imageresize/imageresizeui~ImageResizeOption>} options The resize options.
+	 * @param {module:image/imageresize/imageresizecommand~ImageResizeCommand} command A resize image command.
+	 *
+	 * @returns {module:utils/collection~Collection} definitions
+	*/
```
Comment (oleq) [inline packages/ckeditor5-image/src/imageresize/imageresizeui.js:244]: Wrong indentation.

File: packages/ckeditor5-image/src/imageresize/imageresizeui.js [hunk 6]
```diff
@@ -105,13 +163,11 @@ export default class ImageResizeUI extends Plugin {
 	 * @private
 	 *
 	 * @param {Array.<module:image/imageresize/imageresizeui~ImageResizeOption>} options An array of the configured options.
-	 * @param {String} unit A resize unit.
 	 */
-	_addDropdown( options, unit ) {
+	_addDropdown( options ) {
```
Comment (oleq) [inline packages/ckeditor5-image/src/imageresize/imageresizeui.js:167]: `_registerImageResizeDropdown`?

File: packages/ckeditor5-image/src/imageresize/imageresizeui.js [hunk 7]
```diff
@@ -66,22 +89,57 @@ export default class ImageResizeUI extends Plugin {
 	 * @private
 	 *
 	 * @param {module:image/imageresize/imageresizeui~ImageResizeOption} resizeOption A model of resize option.
-	 * @param {String} unit A resize unit.
 	 */
-	_addButton( { name, label, value }, unit ) {
+	_addButton( option ) {
```
Comment (oleq) [inline packages/ckeditor5-image/src/imageresize/imageresizeui.js:93]: `_registerImageResizeButton`?

File: packages/ckeditor5-image/src/imageresize/imageresizeui.js [hunk 8]
```diff
@@ -66,22 +89,57 @@ export default class ImageResizeUI extends Plugin {
...
+					editor,
+					option
+				);
+			}
+
 			button.set( {
-				label: t( label ),
-				withText: true,
-				tooltip: parsedValue ? t( 'Resize image to' ) + ' ' + parsedValue : t( 'Resize image to the original size' ),
+				// Uses `label` property for setting the more verbose text (from tooltip) for ARIA purpose.
```
Comment (oleq) [inline packages/ckeditor5-image/src/imageresize/imageresizeui.js:139]:
```suggestion
				// Use the `label` property for a verbose description for ARIA purposes.
```

File: packages/ckeditor5-image/src/imageresize/imageresizeui.js [hunk 9]
```diff
@@ -146,33 +204,69 @@ export default class ImageResizeUI extends Plugin {
...
+	 * A helper function for parsing resize options definitions.
+	 *
+	 * @private
+	 *
+	 * @param {Array.<module:image/imageresize/imageresizeui~ImageResizeOption>} options The resize options.
+	 * @param {module:image/imageresize/imageresizecommand~ImageResizeCommand} command A resize image command.
+	 *
+	 * @returns {module:utils/collection~Collection} definitions
+	*/
+	_prepareListDefinitions( options, command ) {
```
Comment (oleq) [inline packages/ckeditor5-image/src/imageresize/imageresizeui.js:245]: `_getResizeDropdownListItemDefinitions`?

File: packages/ckeditor5-image/src/imageresize/imageresizeui.js [hunk 10]
```diff
@@ -146,33 +204,69 @@ export default class ImageResizeUI extends Plugin {
...
+				return t( 'Resize image to the original size' );
+			}
+		} else {
+			return option.value ? option.value + this._resizeUnit : t( 'Original' );
+		}
+	}
 
-		const commandCallback = setOptionOn( parsedValue );
+	/**
+	 * A helper function for parsing resize options definitions.
```
Comment (oleq) [inline packages/ckeditor5-image/src/imageresize/imageresizeui.js:236]:
```suggestion
	 * A helper function that parses resize options and returns definitions ready for use in a drop–down.
```

File: packages/ckeditor5-image/src/imageresize/imageresizeui.js [hunk 11]
```diff
@@ -194,6 +288,7 @@ function setOptionOn( value ) {
  *
  * @property {String} resizeOption.name A name of the option used for creating a component.
  * You refer to that name later in the {@link module:image/image~ImageConfig#toolbar}.
- * @property {String} resizeOption.label A label to be displayed with a button.
  * @property {String} resizeOption.value A value of a resize option. `null` value is for resetting an image to its original size.
+ * @property {String} resizeOptions.icon A value of the available icon sizes (`small`, `medium`, `large`, `original`).
```
Comment (oleq) [inline packages/ckeditor5-image/src/imageresize/imageresizeui.js:292]: It looks like this property is also optional because it is not required when the dropdown is created. It makes sense only when individual buttons are created (if I understand the code).

File: packages/ckeditor5-image/tests/imageresize/imageresizeui.js [hunk 12]
```diff
@@ -82,29 +88,28 @@ describe( 'ImageResizeUI', () => {
 		} );
 	} );
 
+	it( 'should create `_resizeUnit`', () => {
+		const unit = plugin._resizeUnit;
+
+		expect( unit ).to.equal( '%' );
+	} );
+
 	describe( 'init()', () => {
```
Comment (oleq) [inline packages/ckeditor5-image/tests/imageresize/imageresizeui.js:97]: I'm confused by these `init()` tests:

*   "should have set "%" resize unit if not defined" asserts the value of the **command** which has nothing to do with `init()` or the `ImageResizeUI` plugin.
*   "should have set "px" resize unit" same as above, what does `expect( command.value.width.includes( 'px' ) ).to.be.true;` have to do with `ImageResizeUI`?

Shouldn't these tests be somewhere among `ImageResizeCommand` tests?

At the same time:

*   "should be disabled when command is disabled" should be under `init()`

File: packages/ckeditor5-image/theme/imageresize.css [hunk 13]
```diff
@@ -23,3 +23,11 @@
 		display: block;
 	}
 }
+
+[dir="ltr"] .ck.ck-button.ck-button_with-text.ck-resize-image-button .ck-button__icon {
+	margin-right: 6px;
```
Comment (oleq) [inline packages/ckeditor5-image/theme/imageresize.css:28]: Can we use `--ck-spacing-*` for compatibility here?

File: packages/ckeditor5-image/src/imageresize/imageresizeui.js [hunk 14]
```diff
@@ -66,22 +89,57 @@ export default class ImageResizeUI extends Plugin {
...
+					case 'large':
+						return iconLarge;
+					default:
+						return null;
+				}
+			};
+
+			if ( !userIcon() ) {
+				/**
+				* Setting {@link module:image/image~ImageConfig#resizeOptions `config.image.resizeOptions`} for the standalone buttons,
```
[the remark this answers is not in the document]
Comment (panr) [reply packages/ckeditor5-image/src/imageresize/imageresizeui.js:112]: Hmm... shouldn't it follow the `CKEditorError()` indentation?

**Comment by panr** [comment]
Phew... I fixed them all 😅

**Comment by oleq** [comment]
@ckeditor/qa-team Can you take a look at this PR? Thanks 🙏

**Comment by panr** [comment]
> Localization entries are missing.
> Missing changelog entry for ckeditor5-core.
> Docs are bleeding (CKEditorError: imageresizeui-missing-icon) 
> http://fake.ckeditor.com/ck/5/ckeditor5/build/docs/ckeditor5/latest/features/image.html#resizing-images.

Fixed it.

**Comment by FilipTokarski** [comment]
I just reported a bug with jumping blue border, which occurs also on master ( and since very long time ), but it gets much more visible when using those new buttons for resizing.  
  
Original issue: https://github.com/ckeditor/ckeditor5/issues/7633  
  
Screencast:  
  
16_image_5

**Comment by FilipTokarski** [comment]
Apart from what I mentioned above, everything seems to be ok 👍
